pandas/_libs/tslibs/offsets.pyx

```diff
@@ -206,3 +272,109 @@ class ApplyTypeError(TypeError):
 # TODO: unused.  remove?
 class CacheableOffset(object):
     _cacheable = True
+
+
+class BeginMixin(object):
+    # helper for vectorized offsets
+
+    def _beg_apply_index(self, i, freq):
+        """Offsets index to beginning of Period frequency"""
+
+        off = i.to_perioddelta('D')
+
+        base, mult = get_freq_code(freq)
+        base_period = i.to_period(base)
+        if self.n <= 0:
+            # when subtracting, dates on start roll to prior
+            roll = np.where(base_period.to_timestamp() == i - off,
+                            self.n, self.n + 1)
+        else:
+            roll = self.n
+
+        base = (base_period + roll).to_timestamp()
+        return base + off
+
+
+class EndMixin(object):
+    # helper for vectorized offsets
+
+    def _end_apply_index(self, i, freq):
+        """Offsets index to end of Period frequency"""
+
+        off = i.to_perioddelta('D')
+
+        base, mult = get_freq_code(freq)
+        base_period = i.to_period(base)
+        if self.n > 0:
+            # when adding, dates on end roll to next
+            roll = np.where(base_period.to_timestamp(how='end') == i - off,
+                            self.n, self.n - 1)
+        else:
+            roll = self.n
+
+        base = (base_period + roll).to_timestamp(how='end')
+        return base + off
+
+
+# ---------------------------------------------------------------------
+# Base Classes
+
+class _BaseOffset(object):
```

Comment:
why are you creating a base class here? what is the purpose?

Reply:
IOW why not simply have 1 Base class (and not a _BaseOffset and a BaseOffset)

Reply:
See comments about remaining cython/pickle issues.

You're absolutely right that in its current form having two separate classes accomplishes nothing.  The idea is that `_BaseOffset` should be a `cdef class`, while `BaseOffset` should be python class.  (`__rfoo__` methods do not play nicely with cython classes).

Reply:
ok that is fine.

Reply:
I would probably leave this as a class for the moment. I am not convinced this actually needs to be a full c-extension class (e.g. its not like we are inheriting from a python c-class here). I don't see the benefit and it has added complexity.

Reply:
The main reason is to achieve immutability.  That's the big roadblock between us and making `__eq__`, '__ne__`, `__hash__` performance not-awful.  (There's an issue somewhere about "scalar types immutable" or something like that)

Reply:
ok